### ISSUE TITLE
Guidance around unit testing where `SyncArbiter` is involved

### ISSUE THREAD

**Issue by brandur** [actix/actix-web#125]
Hi there,

I'm writing an `actix-web`-based application, and was trying to follow the [guide on building unit tests for endpoints](https://actix.github.io/actix-web/guide/qs_8.html#unit-tests). The complication though is that I'm using a `SyncArbiter` in my application [very much like the Diesel example](https://github.com/actix/actix-web/blob/master/examples/diesel/src/main.rs) that's part of this project.

Testing endpoints that don't need a `SyncArbiter` worked fine, but when invoking one that did, I ran into a problem where messages to the arbiter would just hang. I assume that's because I don't have an `actix::System` running, but I've tried a variety of different ways to get this working, and so far have come up with nothing.

`actix::System.run` blocks, so I tried spawning one in its own thread:

``` rust
        let _join = ::std::thread::spawn(move || {
            let system = actix::System::new("podcore-api-test");
            system.run();
        });
```

But that still produces the message `Arbiter is not running` when I try to `start` a `SyncArbiter`. I tried something like this, but it had the same problem:

``` rust
        let _join = ::std::thread::spawn(move || {
            let system = actix::System::new("podcore-api-test");
            let _ = actix::Arbiter::new("arbiter");
            system.run();
        });
```

I tried `TestServer` and a few other things, but nothing with success.

The fundamental problem is that I don't really have a good understanding of what services I'm supposed to start myself, and how I'm supposed to run them (i.e., threaded or otherwise). I realized through trial and error that there's some kind of globals magic going on (and confirmed by the various `thread_local!` statics in the source), but I don't really get how it's all supposed to interact. I tried to look for prior art in the `actix-web` test suite, but so far haven't come up with anything usable.

Could I get some guidance on what exactly I'm supposed to do to get this working? I've opened an issue because I think this could be a good addition to the user guide — the current state is the testing article is useful, but not complete, and I don't think I'll be the only person confused by this sort of thing. Thanks!

I should also note that I know my implementation works fine already because I'm able to start it and use it with cURL. It's just the testing environment that I'm having trouble with.

**Comment by fafhrd91** [comment]
you have to start SyncActor from test config factory 

```rust
let mut srv = TestServer::new(|app| {
   let addr = SyncArbiter::start(3, || {
        DbExecutor(SqliteConnection::establish("test.db").unwrap())
    });

   app.handler(my_handler)
});
```

or `TestServer::with_state()`

**Comment by fafhrd91** [comment]
I think this example is better.

```rust

struct MyState {addr: Addr<..., DbExecutor>};

let mut srv = TestServer::with_factory(|| {
   let addr = SyncArbiter::start(3, || {
        DbExecutor(SqliteConnection::establish("test.db").unwrap())
    });

    let state = MyState{addr: addr};

    let mut app = TestApp::new(state);
    app.handler(my_handler);
    app
});
```

I will add documentation for this and will add guide section

**Comment by brandur** [comment]
Thanks @fafhrd91!

I hate to hit you with a follow up, but a few more points/questions:

1. I _think_ that [`TestApp::new`](https://github.com/actix/actix-web/blob/master/src/test.rs#L228) is private.
2. Does `TestApp` support an async handler, or is it mostly for use with sync?
3. Is there a way to accomplish something similar using the unit-style test where you invoke the handler directly, or would that be impossible when you need to start invoking other actors? e.g., like this example from the docs:

``` rust
    let resp = TestRequest::with_header("content-type", "text/plain")
        .run(index)
        .unwrap();
    assert_eq!(resp.status(), StatusCode::OK);
```

Thanks again.

**Comment by fafhrd91** [comment]
1. ok, you need `with_state()` method

```rust
let mut srv = TestServer::with_state(
    || {
        let addr = SyncArbiter::start(3, || {
            DbExecutor(SqliteConnection::establish("test.db").unwrap())
        });
        MyState{addr: addr}
   },
   |app| app.handler(my_handler) 
);
```

2. you can use `Box<Future<Item=HttpResponse, Error=Error>>` as `Handler` for `app.handler()`
3. right now you need to create `System` and create state. then you can use 
    [SystemRunner::run_until_complete](https://actix.github.io/actix/actix/struct.SystemRunner.html#method.run_until_complete) to run async handler.

**Comment by brandur** [comment]
@fafhrd91 Thanks! I'm not quite out of the woods yet, but managed to get quite a bit further along (and all the way to a working test) with this advice.

**Comment by fafhrd91** [comment]
I added test server builder https://actix.github.io/actix-web/actix_web/test/struct.TestServer.html#method.build_with_state

**Comment by brandur** [comment]
@fafhrd91 Wow, I don't know how you knew, but `build_with_state` turned out to be exactly what I needed for a well-abstracted test suite. Here's what I ended up with (as an example of roughly what it looks like):

``` rust
    #[test]
    fn test_handler_graphql_get() {
        let bootstrap = TestBootstrap::new();
        let mut server = bootstrap.server_builder.start(|app| {
            app.middleware(middleware::log_initializer::Middleware)
                .handler(handler_graphql_get)
        });

        let req = server
            .client(
                Method::GET,
                format!(
                    "/?query={}",
                    percent_encode(b"{podcast{id}}", DEFAULT_ENCODE_SET)
                ).as_str(),
            )
            .finish()
            .unwrap();
        let resp = server.execute(req.send()).unwrap();

        assert_eq!(resp.status(), StatusCode::OK);

        let bytes: Bytes = resp.body().wait().unwrap();
        let value: serde_json::Value = serde_json::from_slice(bytes.as_ref()).unwrap();

        assert_eq!(json!({"data": {"podcast": []}}), value);
    }
```

**Comment by fafhrd91** [comment]
@brandur `build_with_state` is released as part of 0.4.10 version

**Comment by brandur** [comment]
> @brandur build_with_state is released as part of 0.4.10 version

Excellent! Thanks.

**Comment by fafhrd91** [comment]
added guide info
